File: app/javascript/channels/wax_peer_channel.js
import consumer from "./consumer"

consumer.subscriptions.create("WaxPeerChannel", {
  connected() {
    // Called when the subscription is ready for use on the server
    console.log("Connected to Waxpeer live events channel")
  },

  disconnected() {
    // Called when the subscription has been terminated by the server
  },

  received(data) {
    // Called when there's incoming data on the websocket for this channel
  }
});

File: app/channels/wax_peer_channel.rb
class WaxPeerChannel < ApplicationCable::Channel
  def subscribed
    stream_from "waxpeer_live_data_channel"

    get_waxpeer_live_events
  end

  def unsubscribed
    # Any cleanup needed when channel is unsubscribed
  end

  def get_waxpeer_live_events
    EM.run do
      ws = Faye::WebSocket::Client.new('wss://wssex.waxpeer.com')

      ws.on :open do |event|
        puts "Connected to waxpeer socket"

        # Schedule a recurring task to send a message every 25 seconds to keep the connection live
        EM.add_periodic_timer(25) do
          puts "Sending ping message"
          ws.send("ping")
          puts "Ping message sent"
        end
      end
    
      ws.on :message do |event|
        p [:message, event.data]
      end
    
      ws.on :close do |event|
        puts "Disconnected to waxpeer socket"
      end
    end
  end
end
